smart_open/ftp.py

```diff
@@ -0,0 +1,94 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2019 Radim Rehurek <me@radimrehurek.com>
+#
+# This code is distributed under the terms and conditions
+# from the MIT License (MIT).
+#
+
+"""Implements I/O streams over FTP.
+"""
+
+import getpass
+import logging
+import urllib.parse
+import smart_open.utils
+
+logger = logging.getLogger(__name__)
+
+SCHEME = "ftp"
+
+"""Supported URL schemes."""
+
+DEFAULT_PORT = 21
+
+URI_EXAMPLES = (
+    'ftp://username@host/path/file',
+    'ftp://username:password@host/path/file',
+    'ftp://username:password@host:port/path/file'
+)
+
+
+def _unquote(text):
+    return text and urllib.parse.unquote(text)
+
+
+def parse_uri(uri_as_string):
+    split_uri = urllib.parse.urlsplit(uri_as_string)
+    assert split_uri.scheme in SCHEME
+    return dict(
+        scheme=split_uri.scheme,
+        uri_path=_unquote(split_uri.path),
+        user=_unquote(split_uri.username),
+        host=split_uri.hostname,
+        port=int(split_uri.port or DEFAULT_PORT),
+        password=_unquote(split_uri.password),
+    )
+
+
+def open_uri(uri, mode, transport_params):
+    smart_open.utils.check_kwargs(open, transport_params)
+    parsed_uri = parse_uri(uri)
+    uri_path = parsed_uri.pop('uri_path')
+    parsed_uri.pop('scheme')
+    return open(uri_path, mode, transport_params=transport_params, **parsed_uri)
+
+
+def convert_transport_params_to_args(transport_params):
+    supported_keywords = ['keyfile', 'certfile', 'context', 'timeout', 'source_address', 'encoding']
+    unsupported_keywords = [k for k in transport_params if k not in supported_keywords]
+    kwargs = {k: v for (k, v) in transport_params.items() if k in supported_keywords}
+
+    if unsupported_keywords:
+        logger.warning('ignoring unsupported ftp keyword arguments: %r', unsupported_keywords)
+
+    return kwargs
+
+
+def _connect(hostname, username, port, password, transport_params):
+    try:
+        from ftplib import FTP_TLS
+    except ImportError:
+        raise Exception("Error: Unable to import FTP_TLS from ftplib library")
+    kwargs = convert_transport_params_to_args(transport_params)
+    ftp = FTP_TLS(**kwargs)
+    try:
+        ftp.connect(hostname, port)
+        ftp.sendcmd(f'USER {username}')
+        ftp.sendcmd(f'PASS {password}')
```

Comment:
I imagine many users who want to use TLS would like a secure connection, which would require adding `ftp.prot_p()`
https://docs.python.org/3/library/ftplib.html#ftplib.FTP_TLS.prot_p

Reply:
This is a good point. For now, I am adding functionality for handling reading from a pure FTP server, but I am planning on making a future PR for handling reading from an FTPS server.